src/planwise/component/facilities.clj

```diff
@@ -39,17 +39,50 @@
 ;; ----------------------------------------------------------------------
 ;; Service functions
 
-(defn insert-facilities! [service dataset-id facilities]
+(defn insert-facilities!
+  "Upserts the facilities, identifying them by dataset-id and site-id,
+   returning a hash of state into list of ids, where state is one of
+   :existing, :updated, :moved or :new."
+  [service dataset-id facilities]
   (jdbc/with-db-transaction [tx (get-db service)]
-    (delete-facilities-in-dataset-by-site-id! tx {:dataset-id dataset-id, :site-ids (map :site-id facilities)})
-    (reduce (fn [ids facility]
-              (let [result (insert-facility! tx (assoc facility :dataset-id dataset-id))]
-                (conj ids (:id result))))
-            []
-            facilities)))
-
-(defn destroy-facilities! [service dataset-id]
-  (delete-facilities-in-dataset! (get-db service) {:dataset-id dataset-id}))
+    (let [updateable-keys [:name :type-id :lat :lon]
+          site-ids (map :site-id facilities)
+          existing-facilities (when (seq site-ids)
+                                (facilities-in-dataset-by-criteria tx
+                                  {:dataset-id dataset-id
+                                   :criteria (criteria-snip {:site-ids site-ids})}))]
+      (->> facilities
+        (map  (fn [facility]
+                (if-let [{id :id, :as existing} (some-> (filter #(= (:site-id %) (:site-id facility))
+                                                         existing-facilities)
+                                                  (first)
```

Comment:
Better to use `reduce` similar to https://github.com/instedd/planwise/blob/master/src/planwise/client/utils.cljs#L84